Add phi2 to wormhole

BenraouaneSoufiane opened the pull request:
### Ticket
[Link to Github Issue
](https://github.com/tenstorrent/tt-metal/issues/19587)

### Problem description
Phi2 not available in the demos models

### What's changed
I used the standard approach for compiling BERT models which comes with the starter template (tt_transformers)
Adopt phi2 weights with the existing weights
For non existing weight (ffn_norm), I used dummy weight for rmsnorm (prefer keep it rather than use layernorm without generating dummy weight)
Pad cos to full dim with ones (& sin with zeros) from 80 to 96 to be ready for rottary embedding (in prefill)
Edit the existing MLP & apply which comes with PhiForCasualllm (only two weights fc1 & fc2 instead three up, down, gate)
Slice the att out back to 80 to be ready for the concatinating
Slice the first 1280 of attn out to match the linear with the dense layer (in prefill)
Change the shard of dense layer from (1280, 224) to (1280, 256)
change the num_heads & num_kv_heads from 32, 32 to 32, 4 in decode
I used projector instead slice concern the last dim of attn out to match the dim of the dense layer in matmul in the decode step

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

Commit:
Add Phi 2

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
fix missed ffn_norm

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
shard projected attn output & densed

Commit:
remove tracing (not works)

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
create phi2 for wormhole

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
fix num_cores & enhance attn output shape

Commit:
Update PERF.md

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'main' into main

yieldthought commented:
Results look good, but we can't accept the PR in this form. TT-Transformers is general-purpose for supporting a wide range of model families, including Llama, Mistral, Qwen and Phi 4. We can't accept a duplicate of 238 files and 70k lines of code as a PR to add Phi 2 support.

What we are looking for in this case is a minimal patch to `models/tt_transformers` that adds Phi 2 support to the existing code without breaking existing model support.

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Rest rmsnorm.py

Remove set ffn_norm (missed key for phi2) with dummy weight which break the tt_transformers & making it not standard (works fine for all models), will moved to load_checkpoints.py

Commit:
Delete models/phi2 directory

Commit:
Update load_checkpoints.py

Set missed weight "ffn_norm" with dummy weight (does not affect the final result) keeping rmsnorm instead using layernorm without dummy weight

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Update simple_text_demo.py

Set the HF_MODEL to microsoft/phi-2, so the user not required to init it from his bash or from command line when use this model

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Update README.md

Created the phi 2 version of the intro

Commit:
Update README.md

Fix the wrong command

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

BenraouaneSoufiane commented:
> Results look good, but we can't accept the PR in this form. TT-Transformers is general-purpose for supporting a wide range of model families, including Llama, Mistral, Qwen and Phi 4. We can't accept a duplicate of 238 files and 70k lines of code as a PR to add Phi 2 support.
> 
> What we are looking for in this case is a minimal patch to `models/tt_transformers` that adds Phi 2 support to the existing code without breaking existing model support.
> 
> Users should be able to run `HF_MODEL=microsoft/phi-2 pytest models/tt_transformers/demo/simple_text_demo.py` and have it work out of the box, just as they can with Qwen, Llama, Mistral, Phi 4 and similar models today.

Updated, @yieldthought can you take a look

Commit:
Merge branch 'tenstorrent:main' into main

Commit:
Merge branch 'tenstorrent:main' into main

BenraouaneSoufiane commented:
Can someone review changes? @cglagovichTT @mtairum @uaydonat

Commit:
Merge branch 'tenstorrent:main' into main

yieldthought commented:
`models/demos/phi2` still appears to be a clone to `models/tt_transformers`, adding 120 files and 35,000 lines of code. We cannot accept a PR that duplicates all this code. For this to be accepted, models/demos/phi2 should be deleted entirely and `models/tt_transformers` modified minimally to allow it to correctly execute Phi2 models in addition to the other models it can already execute.

yieldthought reviewed (changes requested):
models/demos/phi2 still appears to be a clone of models/tt_transformers, adding 120 files and 35,000 lines of code. We cannot accept a PR that duplicates all this code. For this to be accepted, models/demos/phi2 should be deleted entirely and models/tt_transformers modified minimally to allow it to correctly execute Phi2 models in addition to the other models it can already execute.

BenraouaneSoufiane commented:
> models/demos/phi2 still appears to be a clone of models/tt_transformers, adding 120 files and 35,000 lines of code. We cannot accept a PR that duplicates all this code. For this to be accepted, models/demos/phi2 should be deleted entirely and models/tt_transformers modified minimally to allow it to correctly execute Phi2 models in addition to the other models it can already execute.

The issue is not add phi2 support to the tt_transformers, the issue is to compile & run phi2 on/for wormhole

Commit:
Merge branch 'tenstorrent:main' into main

BenraouaneSoufiane commented:
@yieldthought there are no minimally/not minimally although I read "model bring up", following the standard BERT implementation & the INFERENCE IS WORKING + THE IMPLEMENTATION NOT SPECIFIED IN THE ISSUE

Commit:
Merge branch 'tenstorrent:main' into main

BenraouaneSoufiane commented:
@yieldthought are you ok??? what do you expect the implementation to be? The BERT methods are fixed (attention, mlp, rottary embedding, sdpa) specially it achieves now the ideal throughput

yieldthought commented:
@BenraouaneSoufiane I'm not determining whether or not you met the bounty conditions, you should have that conversation with whoever your contact person is. As someone responsible for the tt-transformers and models code I will not be adding 35,000 lines of duplicated code to our repo to add support for a single model.

Clearly there has been poor communication in this process and I sympathise for that and am not placing any blame. Independent of that: it's a clear reality that duplicating this much code would be harmful to our ongoing codebase and I am not going to approve that.

BenraouaneSoufiane commented:
> @BenraouaneSoufiane I'm not determining whether or not you met the bounty conditions, you should have that conversation with whoever your contact person is. As someone responsible for the tt-transformers and models code I will not be adding 35,000 lines of duplicated code to our repo to add support for a single model.
> 
> Clearly there has been poor communication in this process and I sympathise for that and am not placing any blame. Independent of that: it's a clear reality that duplicating this much code would be harmful to our ongoing codebase and I am not going to approve that.

There is "model bring up" that I should or anyone should use for his model as per the official doc & this is sufficient, the "minimally" does not specified "how will be" in the bounty & I'm not going to start from zero because of the same things mentionned in the previous comment, your role I think is to validate throughput, it seems you don't understand my previous comment

Shubhamsaboo commented:
@BenraouaneSoufiane stop with your unprofessional comments if you like to get paid and apologise to our engineers for talking like that. We won't tolerate this behaviour. 

When it comes to your PR -- it's not very useful for us. The implementation is a clone of tt-transofrmers library so we won't be merging this PR. 

Said that, since you have put the work and the model can run. We will pay you the bounty and close the issue if your behaviour is in line. BUT this is my final warming to maintain public decorum.
